analytics.go

```diff
@@ -303,45 +308,116 @@ func (c *client) sendAsync(msgs []message, wg *sync.WaitGroup, ex *executor) {
 	}
 }
 
-// Send batch request.
-func (c *client) send(msgs []message) {
+//Split based on Anonymous ID
+func (c *client) getNodePayload(msgs []message) map[int][]message {
+	nodePayload := make(map[int][]message)
+	totalNodes := c.totalNodes
+	for _, msg := range msgs {
+		anonymousId := gjson.GetBytes(msg.json, "anonymousId").String()
+		hashInt := crc32.ChecksumIEEE([]byte(anonymousId))
+		nodePayload[int(hashInt)%totalNodes] = append(nodePayload[int(hashInt)%totalNodes], msg)
+	}
+	return nodePayload
+}
+
+func (c *client) getRevisedMsgs(nodePayload map[int][]message, startFrom int) []message {
+	msgs := make([]message, 0)
+	for k, v := range nodePayload {
+		if k >= startFrom {
+			for _, msg := range v {
+				msgs = append(msgs, msg)
+			}
+		}
+	}
+	return msgs
+}
+
+func (c *client) setNodeCount() {
 	const attempts = 10
+	for i := 0; i < attempts; i++ {
+		url := c.Endpoint + "/cluster-info"
+		req, err := http.NewRequest("GET", url, bytes.NewReader([]byte{}))
+		if err != nil {
+			c.errorf("creating request - %s", err)
+			continue
+		}
+
+		req.Header.Add("User-Agent", "analytics-go (version: "+Version+")")
+		req.SetBasicAuth(c.key, "")
+
+		res, err := c.http.Do(req)
+
+		if err != nil {
+			c.errorf("sending request - %s", err)
+			continue
+		}
+		if res.StatusCode == 200 {
+			body, err := ioutil.ReadAll(res.Body)
+			if err == nil {
+				c.totalNodes = int(gjson.GetBytes(body, "nodeCount").Int())
+				return
+			} else {
+				res.Body.Close()
+			}
+		}
+		res.Body.Close()
+	}
+	return
+}
 
-	b, err := json.Marshal(batch{
+func (c *client) getMarshalled(msgs []message) ([]byte, error) {
+	nodeBatch, err := json.Marshal(batch{
 		MessageId: c.uid(),
 		SentAt:    c.now(),
 		Messages:  msgs,
 		Context:   c.DefaultContext,
 	})
+	return nodeBatch, err
+}
 
-	if err != nil {
-		c.errorf("marshalling messages - %s", err)
-		c.notifyFailure(msgs, err)
-		return
-	}
-
-	for i := 0; i != attempts; i++ {
-		if err = c.upload(b); err == nil {
-			c.notifySuccess(msgs)
-			return
-		}
-
-		// Wait for either a retry timeout or the client to be closed.
-		select {
-		case <-time.After(c.RetryAfter(i)):
-		case <-c.quit:
-			c.errorf("%d messages dropped because they failed to be sent and the client was closed", len(msgs))
-			c.notifyFailure(msgs, err)
-			return
+// Send batch request.
+func (c *client) send(msgs []message) {
+	const attempts = 20
```

Comment:
Why have we changed the earlier behaviour ? I think, previously it was 10

Reply:
This was increased because attempts of 10 would mean we would be retrying for around 1 minute before we drop the data.By increasing the attempts to 20 , we would be giving more time for the SDK to able to send the data to respective kubernetes node while it gets ready to accept the data